### ISSUE TITLE
Consider Using `clippy` in CI

### ISSUE THREAD

**Issue by jalil-salame** [wez/wezterm#3286]
**Is your feature request related to a problem? Please describe.**
I've been applying some simple fixes from clippy as PRs, many of these would be caught and fixed in the original PRs if `clippy` was run on CI.

**Describe the solution you'd like**
Add a Github Workflow to the CI that runs clippy: https://doc.rust-lang.org/clippy/continuous_integration/github_actions.html.

I would suggest starting by allowing all of the current warning lints and merge PRs that incrementally remove those flags.

**Describe alternatives you've considered**
Not using clippy(?)

**Additional context**
I checked what would be left once all of my PRs are merged and clippy found two potential bugs:

```console
error: using `clone` on a double-reference; this will copy the reference of type `&[&str]` instead of cloning the inner type
   --> wezterm-dynamic/src/error.rs:139:39
    |
139 | ...                   possible: possible.clone(),
    |                                 ^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_double_ref
    = note: `#[deny(clippy::clone_double_ref)]` on by default
help: try dereferencing it
    |
139 |                             possible: &(*possible).clone(),
    |                                       ~~~~~~~~~~~~~~~~~~~~
help: or try being explicit if you are sure, that you want to clone a reference
    |
139 |                             possible: <&[&str]>::clone(possible),
    |                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~

error: this loop never actually loops
   --> wezterm-dynamic/src/error.rs:204:13
    |
204 | /             for err in errors {
205 | |                 return Err(err);
206 | |             }
    | |_____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#never_loop
    = note: `#[deny(clippy::never_loop)]` on by default
help: if you need the first element of the iterator, try writing
    |
204 |             if let Some(err) = errors.into_iter().next() {
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

There have also been some changes which would've been covered by my PRs that have already been merged.

**Comment by jalil-salame** [comment]
I spent some time modifying the `fmt` CI workflow and came up with this:

```yaml
name: clippy

on:
  pull_request:
    branches:
      - main
    paths:
      - "**/*.rs"
      - ".github/workflows/clippy.yml"
  push:
    branches:
      - main
    paths:
      - "**/*.rs"
      - ".github/workflows/clippy.yml"

env:
  RUSTFLAGS: "-Dwarnings"

jobs:
  lint-rust-code:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: "Install Rust"
        uses: actions-rs/toolchain@v1
        with:
          profile: "minimal"
          toolchain: "nightly"
          override: true
          components: rustfmt
        env:
          ACTIONS_ALLOW_UNSECURE_COMMANDS: "true"
      - name: Run Clippy # Allow lints with -A lint_name
        run: |
          source $HOME/.cargo/env
          cargo +nightly clippy --all-features --all-targets -- \
                          -A clippy::map_clone \
                          -A clippy::needless_question_mark \
                          -A clippy::needless_borrow \
                          -A clippy::new_without_default \
                          -A clippy::unused_unit \
                          -A clippy::needless_lifetimes \
                          -A clippy::option_as_ref_deref \
                          -A clippy::wrong_self_convention \
                          -A clippy::box_default \
                          -A clippy::manual_map \
                          -A clippy::redundant_field_names \
                          -A clippy::explicit_auto_deref \
                          -A clippy::collapsible_if \
                          -A clippy::missing_safety_doc \
                          -A clippy::useless_conversion \
                          -A clippy::single_component_path_imports \
                          -A clippy::unit_arg \
                          -A clippy::borrow_deref_ref
```

It allows all of the `warn` lints that are currently tripping clippy up.

**Comment by wez** [comment]
Can that list of allowed lints be placed into a config file?
Most of my development is direct; I don't make PRs for myself.
If I were to run clippy locally, I don't want to have to type in all those exceptions.

**Comment by jalil-salame** [comment]
Sadly the only configuration is crate level:

```rust
/// inside lib.rs or main.rs
#![allow(lint_name)]
```

This would touch many more files.

An alternative would be having a shell script `clippy.sh`, which you could run more easily and calling that from ci.

I am happy to add the `#![allow(lint)]` in a PR if you are willing to review that, but if you don't like a lint, then it would be necessary to dupplicate that line on every crate.

**Comment by jalil-salame** [comment]
There are more lints that need to be ignored, I fixed some warnings and now more are popping out, it seems clippy limits the amount sent.

**Comment by wez** [comment]
I'd rather have a clippy.sh than have it "infect" every crate

**Comment by jalil-salame** [comment]
Closed as wontfix; reason: lov wavlue, see #3289
